src/curl.c

```diff
@@ -539,7 +526,107 @@ static int swupd_curl_get_file_full(const char *url, char *filename,
 		unlink(filename);
 	}
 
-	return err;
+	return -status;
+}
+
+/*
+ * Determines what strategy to use based on the return code:
+ * - do not retry: if the fault indicates that the failure isn't transient or
+ *                 is unlikely to be successful if repeated (i.e. disk full)
+ * - retry immediately: if the specific fault reported is unusual or rare, it
+ *                      might have been caused by unusual circumstances, the
+ *                      same failure is unlikely to be repeated (e.g. corrupt
+ *                      network package)
+ * - retry after a delay: if the fault is caused by a transient fault (like
+ *                        network connectivity)
+ */
+static enum retry_strategy determine_strategy(int ret_code)
+{
+	switch (ret_code) {
+	case DOWNLOAD_STATUS_FORBIDDEN:
+	case DOWNLOAD_STATUS_NOT_FOUND:
+	case DOWNLOAD_STATUS_WRITE_ERROR:
+		return DONT_RETRY;
+	case DOWNLOAD_STATUS_RANGE_ERROR:
+	case DOWNLOAD_STATUS_PARTIAL_FILE:
+		return RETRY_NOW;
+	case DOWNLOAD_STATUS_ERROR:
+	case DOWNLOAD_STATUS_TIMEOUT:
+		return RETRY_WITH_DELAY;
+	default:
+		return RETRY_NOW;
+	}
+}
+
+static int retry_download(const char *url, char *filename, struct curl_file_data *in_memory_file, bool resume_ok)
```

Comment:
I would change the name of this function. It's not just retrying. it's trying and then retrying. First time I read this name I though that this function would be called only when we had a download failed.

Reply:
ok, I named it like that because the implementation is based on the "retry" pattern. Let me think on a different name then....

Reply:
what about _retry_download_loop_ ? or what do you suggest?

Reply:
good

Reply:
done